sound/soc/amd/acp/acp-platform.c

```diff
@@ -0,0 +1,311 @@
+	.period_bytes_min = CAPTURE_MIN_PERIOD_SIZE,
+	.period_bytes_max = CAPTURE_MAX_PERIOD_SIZE,
+	.periods_min = CAPTURE_MIN_NUM_PERIODS,
+	.periods_max = CAPTURE_MAX_NUM_PERIODS,
+};
+
+static irqreturn_t i2s_irq_handler(int irq, void *data)
+{
+	struct acp_dev_data *adata = data;
+	struct acp_stream *stream;
+	u16 i2s_flag;
+	u32 val, i = 0;
+
+	if (!adata)
+		return IRQ_NONE;
+
+	i2s_flag = 0;
+
+	val = acp_readl(adata->acp3x_base + ACP_EXTERNAL_INTR_STAT);
+
+	for (i = 0; i < ACP_MAX_STREAM; i++) {
+		stream = adata->stream[i];
+		if (stream != NULL && (val & stream->irq_bit)) {
+			acp_writel(stream->irq_bit, adata->acp3x_base +
+				   ACP_EXTERNAL_INTR_STAT);
+			snd_pcm_period_elapsed(stream->substream);
+			i2s_flag = 1;
+			break;
+		}
+	}
+
+	if (i2s_flag)
+		return IRQ_HANDLED;
+	else
+		return IRQ_NONE;
+}
+
+static void config_pte_for_stream(struct acp_dev_data *adata, struct acp_stream *stream)
+{
+	u32 pte_reg, pte_size, reg_val;
+	unsigned int offset = 0;
+
+	pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_5;
+	pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_5;
+	stream->reg_offset = 0x02000000;
+
+	/* Group Enable */
+	reg_val = ACP_SRAM_PTE_OFFSET + offset;
+	acp_writel(reg_val | BIT(31), adata->acp3x_base + pte_reg);
+	acp_writel(PAGE_SIZE_4K_ENABLE,  adata->acp3x_base + pte_size);
+
+}
+static void config_acp_dma(struct acp_dev_data *adata, int cpu_id, int size)
+{
+	u16 page_idx;
+	u32 low = 0, high = 0, val = 0;
+	dma_addr_t addr;
+	struct acp_stream *stream;
+	struct snd_pcm_substream *substream;
+	int direction, num_pages;
+	u32 low2, high2, offset2;
+
+	stream = adata->stream[cpu_id];
+	substream = stream->substream;
+	direction = substream->stream;
+
+	addr = substream->dma_buffer.addr;
+	num_pages = (PAGE_ALIGN(size) >> PAGE_SHIFT);
+
+	val = stream->pte_offset;
+
+	for (page_idx = 0; page_idx < num_pages; page_idx++) {
+		/* Load the low address of page int ACP SRAM through SRBM */
+		low = lower_32_bits(addr);
+		high = upper_32_bits(addr);
+		acp_writel(low, adata->acp3x_base + ACP_SCRATCH_REG_0 + val);
+		high |= BIT(31);
+		acp_writel(high, adata->acp3x_base + ACP_SCRATCH_REG_0 + val
+				+ 4);
+
+		low2 = acp_readl(adata->acp3x_base + ACP_SCRATCH_REG_0 + val);
+		high2 = acp_readl(adata->acp3x_base + ACP_SCRATCH_REG_0 + val + 4);
+		offset2 =  ACP_SCRATCH_REG_0 + val;
+		/* Move to next physically contiguous page */
+		val += 8;
+		addr += PAGE_SIZE;
+	}
+}
+
+static int acp_dma_open(struct snd_soc_component *component,
+			  struct snd_pcm_substream *substream)
+{
+	struct snd_pcm_runtime *runtime = substream->runtime;
+	struct snd_soc_pcm_runtime *soc_runtime = asoc_substream_to_rtd(substream);
+	struct snd_soc_dai *cpu_dai = asoc_rtd_to_cpu(soc_runtime, 0);
+	struct device *dev = component->dev;
+	struct acp_dev_data *adata = dev_get_drvdata(dev);
+	int ret = 0;
+	int stream_id = (cpu_dai->driver->id)*2 + substream->stream;
+	struct acp_stream *stream;
+
+	stream = devm_kzalloc(dev, sizeof(*stream), GFP_KERNEL);
```

Comment:
if (!stream)
    return -ENOMEM


Also it's not a good practice to use devm_ at the component level, this should be used only for device-level probe.